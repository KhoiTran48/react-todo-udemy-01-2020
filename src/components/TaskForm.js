import React, { Component } from 'react'
import moment from 'moment'
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';


export default class TaskForm extends Component {
    constructor(props){
        super(props)
        this.state={
            taskId: props.taskId ? props.taskId : "",
            title : props.title ? props.title : "",
            desc : props.desc ? props.desc : "",
            deadline: props.deadline ? moment(props.deadline) : moment() 
        }
    }

    handleInput = (e)=>{
        if(e.target.value){
            this.setState({[e.target.name]: e.target.value})
        }
    }

    handleSubmit = (e) =>{
        let deadline = this.state.deadline.valueOf();
        this.props.handleSubmit({...this.state, deadline})
    }

    handleDate = (deadline) => {
        if(deadline){
            this.setState({deadline})
        }
    }

    render() {
        return (
            <div>
                Title
                <input 
                    type="text" 
                    name="title" 
                    value={this.state.title}
                    onChange={this.handleInput}
                    autoFocus
                    />
                Desc
                <input 
                    type="text" 
                    name="desc" 
                    value={this.state.desc}
                    onChange={this.handleInput}
                    />
                Created At
                <SingleDatePicker
                    date={this.state.deadline} // momentPropTypes.momentObj or null
                    onDateChange={this.handleDate} // PropTypes.func.isRequired
                    focused={this.state.focused} // PropTypes.bool
                    onFocusChange={({ focused }) => this.setState({ focused })} // PropTypes.func.isRequired
                    numberOfMonths={1}
                    />
                <button onClick={this.handleSubmit}>Save</button>
            </div>
        )
    }
}
