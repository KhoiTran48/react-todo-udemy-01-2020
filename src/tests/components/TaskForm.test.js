import React from 'react'
import TaskForm from '../../components/TaskForm'
import {shallow} from 'enzyme'

test("render task form", ()=>{
    const wrapper = shallow(<TaskForm/>)
    expect(wrapper).toMatchSnapshot();
})

test("change state: title", ()=>{
    const title = "New title";
    const wrapper = shallow(<TaskForm/>);
    const event = {
        target: {
            name: "title",
            value: title
        }
    }
    // wrapper.find("input").at(1).simulate("change",{    
    wrapper.find("input[name='title']").simulate("change",event)

    expect(wrapper.state("title")).toBe(title)

})