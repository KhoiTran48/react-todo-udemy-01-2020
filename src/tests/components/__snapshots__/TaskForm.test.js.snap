// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render task form 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TaskForm />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        "Title",
        <input
          autoFocus={true}
          name="title"
          onChange={[Function]}
          type="text"
          value=""
/>,
        "Desc",
        <input
          name="desc"
          onChange={[Function]}
          type="text"
          value=""
/>,
        "Created At",
        <withStyles(SingleDatePicker)
          anchorDirection="left"
          appendToBody={false}
          ariaLabel={undefined}
          block={false}
          calendarInfoPosition="bottom"
          customCloseIcon={null}
          customInputIcon={null}
          date={"1970-01-01T00:00:00.000Z"}
          dayAriaLabelFormat={undefined}
          dayPickerNavigationInlineStyles={null}
          daySize={39}
          disableScroll={false}
          disabled={false}
          displayFormat={[Function]}
          enableOutsideDays={false}
          firstDayOfWeek={null}
          focused={false}
          hideKeyboardShortcutsPanel={false}
          horizontalMargin={0}
          horizontalMonthPadding={13}
          id="date"
          initialVisibleMonth={null}
          inputIconPosition="before"
          isDayBlocked={[Function]}
          isDayHighlighted={[Function]}
          isOutsideRange={[Function]}
          isRTL={false}
          keepFocusOnInput={false}
          keepOpenOnDateSelect={false}
          monthFormat="MMMM YYYY"
          navNext={null}
          navPosition="navPositionTop"
          navPrev={null}
          noBorder={false}
          numberOfMonths={1}
          onClose={[Function]}
          onDateChange={[Function]}
          onFocusChange={[Function]}
          onNextMonthClick={[Function]}
          onPrevMonthClick={[Function]}
          openDirection="down"
          orientation="horizontal"
          phrases={
                    Object {
                              "calendarLabel": "Calendar",
                              "chooseAvailableDate": [Function],
                              "clearDate": "Clear Date",
                              "closeDatePicker": "Close",
                              "dateIsSelected": [Function],
                              "dateIsUnavailable": [Function],
                              "enterKey": "Enter key",
                              "escape": "Escape key",
                              "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
                              "homeEnd": "Home and end keys",
                              "jumpToNextMonth": "Move forward to switch to the next month.",
                              "jumpToPrevMonth": "Move backward to switch to the previous month.",
                              "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
                              "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
                              "keyboardShortcuts": "Keyboard Shortcuts",
                              "leftArrowRightArrow": "Right and left arrow keys",
                              "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
                              "moveFocusByOneMonth": "Switch months.",
                              "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
                              "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
                              "openThisPanel": "Open this panel.",
                              "pageUpPageDown": "page up and page down keys",
                              "questionMark": "Question mark",
                              "returnFocusToInput": "Return to the date input field.",
                              "roleDescription": "datepicker",
                              "selectFocusedDate": "Select the date in focus.",
                              "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
                              "upArrowDownArrow": "up and down arrow keys",
                            }
          }
          placeholder="Date"
          readOnly={false}
          regular={false}
          renderCalendarDay={undefined}
          renderCalendarInfo={null}
          renderDayContents={null}
          renderMonthElement={null}
          renderMonthText={null}
          renderNavNextButton={null}
          renderNavPrevButton={null}
          renderWeekHeaderElement={null}
          reopenPickerOnClearDate={false}
          required={false}
          screenReaderInputMessage=""
          showClearDate={false}
          showDefaultInputIcon={false}
          small={false}
          transitionDuration={undefined}
          verticalHeight={null}
          verticalSpacing={22}
          weekDayFormat="dd"
          withFullScreenPortal={false}
          withPortal={false}
/>,
        <button
          onClick={[Function]}
>
          Save
</button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      "Title",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "autoFocus": true,
          "name": "title",
          "onChange": [Function],
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      "Desc",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "name": "desc",
          "onChange": [Function],
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      "Created At",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "anchorDirection": "left",
          "appendToBody": false,
          "ariaLabel": undefined,
          "block": false,
          "calendarInfoPosition": "bottom",
          "customCloseIcon": null,
          "customInputIcon": null,
          "date": "1970-01-01T00:00:00.000Z",
          "dayAriaLabelFormat": undefined,
          "dayPickerNavigationInlineStyles": null,
          "daySize": 39,
          "disableScroll": false,
          "disabled": false,
          "displayFormat": [Function],
          "enableOutsideDays": false,
          "firstDayOfWeek": null,
          "focused": false,
          "hideKeyboardShortcutsPanel": false,
          "horizontalMargin": 0,
          "horizontalMonthPadding": 13,
          "id": "date",
          "initialVisibleMonth": null,
          "inputIconPosition": "before",
          "isDayBlocked": [Function],
          "isDayHighlighted": [Function],
          "isOutsideRange": [Function],
          "isRTL": false,
          "keepFocusOnInput": false,
          "keepOpenOnDateSelect": false,
          "monthFormat": "MMMM YYYY",
          "navNext": null,
          "navPosition": "navPositionTop",
          "navPrev": null,
          "noBorder": false,
          "numberOfMonths": 1,
          "onClose": [Function],
          "onDateChange": [Function],
          "onFocusChange": [Function],
          "onNextMonthClick": [Function],
          "onPrevMonthClick": [Function],
          "openDirection": "down",
          "orientation": "horizontal",
          "phrases": Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsSelected": [Function],
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month.",
            "jumpToPrevMonth": "Move backward to switch to the previous month.",
            "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
            "moveFocusByOneMonth": "Switch months.",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
            "openThisPanel": "Open this panel.",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field.",
            "roleDescription": "datepicker",
            "selectFocusedDate": "Select the date in focus.",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
            "upArrowDownArrow": "up and down arrow keys",
          },
          "placeholder": "Date",
          "readOnly": false,
          "regular": false,
          "renderCalendarDay": undefined,
          "renderCalendarInfo": null,
          "renderDayContents": null,
          "renderMonthElement": null,
          "renderMonthText": null,
          "renderNavNextButton": null,
          "renderNavPrevButton": null,
          "renderWeekHeaderElement": null,
          "reopenPickerOnClearDate": false,
          "required": false,
          "screenReaderInputMessage": "",
          "showClearDate": false,
          "showDefaultInputIcon": false,
          "small": false,
          "transitionDuration": undefined,
          "verticalHeight": null,
          "verticalSpacing": 22,
          "weekDayFormat": "dd",
          "withFullScreenPortal": false,
          "withPortal": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Save",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Save",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "Title",
          <input
            autoFocus={true}
            name="title"
            onChange={[Function]}
            type="text"
            value=""
/>,
          "Desc",
          <input
            name="desc"
            onChange={[Function]}
            type="text"
            value=""
/>,
          "Created At",
          <withStyles(SingleDatePicker)
            anchorDirection="left"
            appendToBody={false}
            ariaLabel={undefined}
            block={false}
            calendarInfoPosition="bottom"
            customCloseIcon={null}
            customInputIcon={null}
            date={"1970-01-01T00:00:00.000Z"}
            dayAriaLabelFormat={undefined}
            dayPickerNavigationInlineStyles={null}
            daySize={39}
            disableScroll={false}
            disabled={false}
            displayFormat={[Function]}
            enableOutsideDays={false}
            firstDayOfWeek={null}
            focused={false}
            hideKeyboardShortcutsPanel={false}
            horizontalMargin={0}
            horizontalMonthPadding={13}
            id="date"
            initialVisibleMonth={null}
            inputIconPosition="before"
            isDayBlocked={[Function]}
            isDayHighlighted={[Function]}
            isOutsideRange={[Function]}
            isRTL={false}
            keepFocusOnInput={false}
            keepOpenOnDateSelect={false}
            monthFormat="MMMM YYYY"
            navNext={null}
            navPosition="navPositionTop"
            navPrev={null}
            noBorder={false}
            numberOfMonths={1}
            onClose={[Function]}
            onDateChange={[Function]}
            onFocusChange={[Function]}
            onNextMonthClick={[Function]}
            onPrevMonthClick={[Function]}
            openDirection="down"
            orientation="horizontal"
            phrases={
                        Object {
                                    "calendarLabel": "Calendar",
                                    "chooseAvailableDate": [Function],
                                    "clearDate": "Clear Date",
                                    "closeDatePicker": "Close",
                                    "dateIsSelected": [Function],
                                    "dateIsUnavailable": [Function],
                                    "enterKey": "Enter key",
                                    "escape": "Escape key",
                                    "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
                                    "homeEnd": "Home and end keys",
                                    "jumpToNextMonth": "Move forward to switch to the next month.",
                                    "jumpToPrevMonth": "Move backward to switch to the previous month.",
                                    "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
                                    "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
                                    "keyboardShortcuts": "Keyboard Shortcuts",
                                    "leftArrowRightArrow": "Right and left arrow keys",
                                    "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
                                    "moveFocusByOneMonth": "Switch months.",
                                    "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
                                    "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
                                    "openThisPanel": "Open this panel.",
                                    "pageUpPageDown": "page up and page down keys",
                                    "questionMark": "Question mark",
                                    "returnFocusToInput": "Return to the date input field.",
                                    "roleDescription": "datepicker",
                                    "selectFocusedDate": "Select the date in focus.",
                                    "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
                                    "upArrowDownArrow": "up and down arrow keys",
                                  }
            }
            placeholder="Date"
            readOnly={false}
            regular={false}
            renderCalendarDay={undefined}
            renderCalendarInfo={null}
            renderDayContents={null}
            renderMonthElement={null}
            renderMonthText={null}
            renderNavNextButton={null}
            renderNavPrevButton={null}
            renderWeekHeaderElement={null}
            reopenPickerOnClearDate={false}
            required={false}
            screenReaderInputMessage=""
            showClearDate={false}
            showDefaultInputIcon={false}
            small={false}
            transitionDuration={undefined}
            verticalHeight={null}
            verticalSpacing={22}
            weekDayFormat="dd"
            withFullScreenPortal={false}
            withPortal={false}
/>,
          <button
            onClick={[Function]}
>
            Save
</button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        "Title",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "autoFocus": true,
            "name": "title",
            "onChange": [Function],
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        "Desc",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "desc",
            "onChange": [Function],
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        "Created At",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "anchorDirection": "left",
            "appendToBody": false,
            "ariaLabel": undefined,
            "block": false,
            "calendarInfoPosition": "bottom",
            "customCloseIcon": null,
            "customInputIcon": null,
            "date": "1970-01-01T00:00:00.000Z",
            "dayAriaLabelFormat": undefined,
            "dayPickerNavigationInlineStyles": null,
            "daySize": 39,
            "disableScroll": false,
            "disabled": false,
            "displayFormat": [Function],
            "enableOutsideDays": false,
            "firstDayOfWeek": null,
            "focused": false,
            "hideKeyboardShortcutsPanel": false,
            "horizontalMargin": 0,
            "horizontalMonthPadding": 13,
            "id": "date",
            "initialVisibleMonth": null,
            "inputIconPosition": "before",
            "isDayBlocked": [Function],
            "isDayHighlighted": [Function],
            "isOutsideRange": [Function],
            "isRTL": false,
            "keepFocusOnInput": false,
            "keepOpenOnDateSelect": false,
            "monthFormat": "MMMM YYYY",
            "navNext": null,
            "navPosition": "navPositionTop",
            "navPrev": null,
            "noBorder": false,
            "numberOfMonths": 1,
            "onClose": [Function],
            "onDateChange": [Function],
            "onFocusChange": [Function],
            "onNextMonthClick": [Function],
            "onPrevMonthClick": [Function],
            "openDirection": "down",
            "orientation": "horizontal",
            "phrases": Object {
              "calendarLabel": "Calendar",
              "chooseAvailableDate": [Function],
              "clearDate": "Clear Date",
              "closeDatePicker": "Close",
              "dateIsSelected": [Function],
              "dateIsUnavailable": [Function],
              "enterKey": "Enter key",
              "escape": "Escape key",
              "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
              "homeEnd": "Home and end keys",
              "jumpToNextMonth": "Move forward to switch to the next month.",
              "jumpToPrevMonth": "Move backward to switch to the previous month.",
              "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
              "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
              "keyboardShortcuts": "Keyboard Shortcuts",
              "leftArrowRightArrow": "Right and left arrow keys",
              "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
              "moveFocusByOneMonth": "Switch months.",
              "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
              "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
              "openThisPanel": "Open this panel.",
              "pageUpPageDown": "page up and page down keys",
              "questionMark": "Question mark",
              "returnFocusToInput": "Return to the date input field.",
              "roleDescription": "datepicker",
              "selectFocusedDate": "Select the date in focus.",
              "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
              "upArrowDownArrow": "up and down arrow keys",
            },
            "placeholder": "Date",
            "readOnly": false,
            "regular": false,
            "renderCalendarDay": undefined,
            "renderCalendarInfo": null,
            "renderDayContents": null,
            "renderMonthElement": null,
            "renderMonthText": null,
            "renderNavNextButton": null,
            "renderNavPrevButton": null,
            "renderWeekHeaderElement": null,
            "reopenPickerOnClearDate": false,
            "required": false,
            "screenReaderInputMessage": "",
            "showClearDate": false,
            "showDefaultInputIcon": false,
            "small": false,
            "transitionDuration": undefined,
            "verticalHeight": null,
            "verticalSpacing": 22,
            "weekDayFormat": "dd",
            "withFullScreenPortal": false,
            "withPortal": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Save",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Save",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
